ruma-events-macros/src/event_content.rs

```diff
@@ -61,65 +147,130 @@ pub fn expand_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_ephemeral_room_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let event_content_impl = expand_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", input.ident);
+        quote! {
+            impl ::ruma_events::EphemeralRoomEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #event_content_impl
 
         impl ::ruma_events::EphemeralRoomEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
 
 /// Create a `RoomEventContent` implementation for a struct.
-pub fn expand_room_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_room_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let event_content_impl = expand_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", input.ident);
+        quote! {
+            impl ::ruma_events::RoomEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #event_content_impl
 
         impl ::ruma_events::RoomEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
 
 /// Create a `MessageEventContent` implementation for a struct
-pub fn expand_message_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_message_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let room_ev_content = expand_room_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", &ident);
+        quote! {
+            impl ::ruma_events::MessageEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #room_ev_content
 
         impl ::ruma_events::MessageEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
 
 /// Create a `StateEventContent` implementation for a struct
-pub fn expand_state_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_state_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = input.ident.clone();
     let room_ev_content = expand_room_event_content(input)?;
 
+    let redacted_marker_trait = if needs_redacted(input) {
+        let ident = quote::format_ident!("Redacted{}", input.ident);
+        quote! {
+            impl ::ruma_events::StateEventContent for #ident { }
+        }
+    } else {
+        TokenStream::new()
+    };
+
     Ok(quote! {
         #room_ev_content
 
         impl ::ruma_events::StateEventContent for #ident { }
+
+        #redacted_marker_trait
     })
 }
+
+fn needs_redacted(input: &DeriveInput) -> bool {
```

Comment:
We are already parsing the arguments elsewhere, right? Can we not give this function `&[EventMeta]` as input?

Reply:
The problem is that all the `expand_*_event_content` functions used in `lib.rs` use the `needs_redacted` function and each of them only have access to the `input`? I could return a `Result<(TokenStream, bool)>` from `expand_event_content` maybe?

Reply:
Hm, in that case let's keep this as-is, it can still be re-visited later.